### DISCUSSION TITLE
Network planning map: help "mapping" attribute values between nodes

### DISCUSSION THREAD

**Discussion by silopolis** [freeplane/freeplane#1664] [Help] [answered]
Hi,

As I wasn't able to write a good title, let me explain...

Scratching my head while making network planning, I came to Freeplane to help design IP addressing plans. CIDR hierarchical structure makes mind maps a good fit for representation.
I then started adding static attributes to nodes, add quickly thought adding a bit of smartness would be nice. A quick search led me to the scripting doc and in a snap I had my first dynamic attributes referencing another node's one.

So I now have the CIDR of my various zones simply calculated, awesome!

Pushing further, I wanted to also get the number of hosts available for each zone dynamically. Found the [IPAddress](https://seancfoley.github.io/IPAddress/) library with which it looks like I could make wonders, but it is far, far above my nonexistent Groovy programming skills :'(

So, the next idea was to create a node with a lookup table, to get the value from another attribute. More precisely:
* I created a node with a [CIDR,HostNum] attribute "table",
* Each other node, representing a network zone, has a CIDR attribute

Finally, the question: how can I add another dynamic attribute to each node, looking up into the table the HostNum value corresponding to its CIDR attribute value?

![image](https://github.com/freeplane/freeplane/assets/414540/7d885b10-4a43-4ca2-a045-af03dfb74e06)

**Comment by euu2021** [comment]
If I understood correctly, this script does what you want, but not dynamically (I manually apply to each node): 
![javaw_EfcQwiFCEe](https://github.com/freeplane/freeplane/assets/77707706/1e2ef394-ad17-47c9-ac39-a4ee16c09e8f)

Script:
```
nodeHostsNum = ID_1580210190 //Id number of the Hosts num node.
cidr = node.attributes.get(0).toString() //This is the CIDR of the selected node.
hostNum = nodeHostsNum.attributes.getFirst(cidr) //This gets the host number that correspondes to the CIDR above.
node.attributes.add("host num", hostNum) //This adds the host number to the selected node.
```

If this is correct, then we can go on to think how to make it happen dynamically.

**Reply by silopolis** [reply]
Hey thank you @euu2021, that was fast!

Yes indeed, that's what I need, plus attribute creation! A step closer to "smart network map nodes"  :heart_eyes: 

Meanwhile I tried to mimic something I knew and a couple of failures later, `=ID_1698907268[node["CIDR"].to.string].to.num` happened to work :)
But I have to create the attribute and the reference by ID looks so ugly! Is there a way to that `nodeHostsNum = ID_1580210190` globally to use it in any attribute expression ? I've tried to add an alias and reference it with N(ID_xxxx) to no luck.

For the attribute creation, I could live with a template node I'd copy and paste, but of course, being able to create a special smart "network" node type that could be defined as default for the map would be freaking cool :P

One thing didn't mention is that I need the reference to be a dynamic link because I want to be able to change the value of the node's CIDR attribute and get its HostNum attribute value updated.

**Reply by i-plasm** [reply, marked as answer]
> One thing didn't mention is that I need the reference to be a dynamic link because I want to be able to change the value of the node's CIDR attribute and get its HostNum attribute value updated.

Using formulas will update the value automatically every time you change the CIDR. I just found you can indeed use an alias, but it needs to be set as a global node. The formula would be `=node.at(":~nodeHostsNum")[node["CIDR"].to.string].to.num`. But it requires that you check "Globally accessible" when assigning an alias to the  **Host num** node:

![setglobalalias](https://github.com/freeplane/freeplane/assets/144830522/0bf04028-1962-4f45-8957-c75e5488fb0c)

Result:
![dynamicatt](https://github.com/freeplane/freeplane/assets/144830522/aa624681-e4f8-4d6b-81ba-4bf1e0abd5f7)

**Reply by euu2021** [reply]
> For the attribute creation, I could live with a template node I'd copy and paste, but of course, being able to create a special smart "network" node type that could be defined as default for the map would be freaking cool :P

You can choose a node style to be default new node. For example: `Format->Style for new nodes->ex1` 

And node styles can have default attributes. You set then at `Format->Manage Styles->Edit styles`. The problem is that those attributes are not assigned automatically as the style is assigned to a node. You need to use `Edit->Node properties->Attributes from style`. To make this happen automatically, it's possible to think in a node change listener. As an example of how a node change listener works, take a look at: https://sourceforge.net/p/freeplane/discussion/758437/thread/e9200db6c3/

Now, about node templates, this script is useful: https://docs.freeplane.org/scripting/Scripts_collection.html#template-chooser---create-nodes-from-templates

**Reply by silopolis** [reply]
hi @i-plasm 

>  I just found you can indeed use an alias, but it needs to be set as a global node. The formula would be =node.at(":~nodeHostsNum")[node["CIDR"].to.string].to.num. But it requires that you check "Globally accessible" when assigning an alias to the Host num node

I had suspected that global accessibility requirement as it's the same for attributes in draw.io :+1: 

Considering I've set a "Cloud" alias on my root node, I tried to replace `=node.parent["CIDR"].to.num + 2` with `=node.at(":~Cloud")[node["CIDR"].to.string].to.num + 2` but I'm getting an error.

Must be doing something wrong :thinking:

**Reply by silopolis** [reply]
@euu2021 those look like nice tracks to follow!
Thanks for sharing :pray:

**Reply by i-plasm** [reply]
Hi @silopolis, is the last formula you shared (https://github.com/freeplane/freeplane/discussions/1664#discussioncomment-8300496) intended for determining the CIDR value, correct? If so, your formula using alias would be `=node.at(":~Cloud")["CIDR"].to.num + 2`. The formula I shared in the previous message was for retrieving the Hostnum value from the "Hosts num" table.

**Reply by silopolis** [reply]
Hello @i-plasm
This is emabarassing, your preceding formula was the good one... Don't ask me what I did to break it :/
Anyway, this is working beautifully and has really teased me to try to make something with that IPAddress library...

Thanks again for your help :pray:
Take care

**Reply by i-plasm** [reply]
>  your preceding formula was the good one...Don't ask me what I did to break it :/

No worries, I usually have to double-check (or more) my stuff, otherwise... :|  Glad it worked 👍

**Reply by i-plasm** [reply]
P.S Forgot to say: In theory, the formula results could go stale/not be updated in real time - although I've never have experienced that. You might want to keep in mind the API methods `evaluateAllFormulas`, `evaluateOutdatedFormulas`, and also the cache itself can be disabled here in the preferences: `Tools -> Preferences -> Plugins -> Formulas: Disable formula evaluation cache`

**Comment by dpolivaev** [comment]
Have you considered using just
Edit->Copy->Copy attributes
Edit->Paste attributes
?

**Reply by silopolis** [reply]
@dpolivaev thanks for the tip :)

I'm actually using it ATM and it's already a great relief. Nevertheless, I'd love to have something more automated and dynamic.

@euu2021 leads may allow defining a default node style with all attributes set with their expressions, and hopefully automatic update of nodes when style is modified... That would be perfect :)
